cpp/oneapi/dal/backend/primitives/distance/test/squared_l2_distance_auxilary_dpc.cpp

```diff
@@ -0,0 +1,240 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <array>
+#include <cmath>
+#include <type_traits>
+
+#include "oneapi/dal/test/engine/common.hpp"
+#include "oneapi/dal/test/engine/fixtures.hpp"
+#include "oneapi/dal/test/engine/dataframe.hpp"
+
+#include "oneapi/dal/table/row_accessor.hpp"
+
+#include "oneapi/dal/backend/primitives/distance/distance.hpp"
+#include "oneapi/dal/backend/primitives/distance/squared_l2_distance_auxilary.hpp"
+
+namespace oneapi::dal::backend::primitives::test {
+
+namespace te = dal::test::engine;
+namespace pr = oneapi::dal::backend::primitives;
+
+using types_to_check = std::tuple<float, double>;
+
+template <typename Float>
+class compute_squared_l2_norms_test_random : public te::float_algo_fixture<Float> {
+public:
+    void generate() {
+        r_count_ = GENERATE(17, 31);
+        c_count_ = GENERATE(3, 13);
+        generate_input();
+    }
+
+    te::table_id get_homogen_table_id() const {
+        return te::table_id::homogen<float_t>();
+    }
+
+    auto output() {
+        return ndarray<Float, 1>::zeros(this->get_queue(), { r_count_ });
+    }
+
+    void generate_input() {
+        const auto input_dataframe =
+            GENERATE_DATAFRAME(te::dataframe_builder{ r_count_, c_count_ }.fill_uniform(-0.2, 0.5));
+        this->input_table_ = input_dataframe.get_table(this->get_homogen_table_id());
+    }
+
+    void squared_l2_norms_check(const ndview<Float, 1>& out, const Float atol = 1.e-3) {
+        for (std::int64_t i = 0; i < r_count_; ++i) {
+            const auto inp_row =
+                row_accessor<const Float>{ input_table_ }.pull(this->get_queue(), { i, i + 1 });
+            Float gtv = 0;
+            for (std::int64_t j = 0; j < c_count_; ++j) {
+                gtv += (inp_row[j] * inp_row[j]);
+            }
+            const Float val = *(out.get_data() + i);
+            const Float diff = gtv - val;
+            CAPTURE(gtv, val, i, r_count_, c_count_);
+            REQUIRE(-atol <= diff);
```

Comment:
Relative tolerance?

Reply:
The result can be close to zero. In such case relative tolerance isn't a good choice